### ISSUE TITLE
Make a clear specification how #undef should be handled in Tables.jl

### ISSUE THREAD

**Issue by bkamins** [JuliaData/Tables.jl#186]
This is a follow-up of https://discourse.julialang.org/t/help-to-check-if-an-element-in-a-tables-jl-is-defined/44110.

@quinnj - can you please add a contract you prefer to have in Tables.jl. I agree with you that we can require that `#undef` should not be passed and can throw an error.

We will have a slight problem with DataFrames.jl when printing with PrettyTables.jl but I think it is a corner case that we can work around.

Maybe @ronisbr can make an explicit check of try-catch for `UndefRefError`?

CC @nalimilan as you might have some opinion here.

**Comment by ronisbr** [comment]
Hi @bkamins !

> Maybe @ronisbr can make an explicit check of try-catch for `UndefRefError`?

Wouldn't it hurt a lot the performance? I will need to `try-catch` every cell before printing.

**Comment by bkamins** [comment]
> I will need to try-catch every cell before printing.

I think that the impact will be negligible. I tried:
```
 @time for i in 1:1000
       try
           println(x[i])
       catch e
           println("undef")
       end
   end
```
with `x` having/not having `#undef` and:
```
 @time for i in 1:1000
       println(x[i])
   end
```
where `x[i]` does not have `#undef` and timings are very similar. Simply printing seems to be much more expensive anyway.

But of course please cross-test this.

**Comment by ronisbr** [comment]
@bkamins I did a test and for a table with 180 `#undef` entries and with 270 elements, the code with `try...catch` is 10x slower than the code with `assigned`. Hence, I decided to do this:

* For normal arrays, just use `isassigned`.
* For tables compliant with Tables.jl API that has column access, just use `isassigned`.
* For tables compliant with Tables.jl API that has only row access, use `try...catch`.

**Comment by bkamins** [comment]
This should be OK (@quinnj - please confirm). Thank you for looking into this.

As a side note - `isassigned` uses `try`-`catch` for general arrays 😄.

Probably for `SentinelArray` and `CategoricalArray` custom `isassigned` method should be added (just like `PooledArray` does)?

CC @nalimilan 

And out of couriosity. Is 10x slower with printing to terminal or to string (and what are the absolute timings)?

**Comment by ronisbr** [comment]
> As a side note - `isassigned` uses `try`-`catch` for general arrays 😄.

This is very, very strange! I have no explanation at all. Maybe something very wrong in my implementation?

```julia
function isassigned(rtable::RowTable, inds...)
    length(inds) != 2 &&
    error("A element of type `RowTable` must be accesses using 2 indices.")

    # Access index.
    i,j = inds[1], inds[2]

    # Get the column name.
    column_name = rtable.column_names[j]

    # Get the element.
    it,~ = iterate(rtable.table, i)

    it == nothing &&
    error("The row `i` does not exist.")

    try
        element = it[column_name]
        return true
    catch e
        if isa(e, UndefRefError)
            return false
        else
            throw(e)
        end
    end
end
```

> And out of couriosity. Is 10x slower with printing to terminal or to string (and what are the absolute timings)?

I tested using `@time` and printing to the `stdout`. Using a table with column access, it took 0.001s (mean) to print. Using a table with row access, it took 0.095s (mean).

**Comment by nalimilan** [comment]
> Probably for `SentinelArray` and `CategoricalArray` custom `isassigned` method should be added (just like `PooledArray` does)?

@bkamins Yes, if that's a bottleneck in some workflows a custom method to check whether the ref is 0 can easily be added.

@ronisbr Maybe the performance difference is due to the fact that `isassigned` is in a separate function? Adding `try... catch` may disable inlining. You can try with an explicit `@inline`.

**Comment by bkamins** [comment]
@ronisbr Just for your reference. The generic implementation of `isassigned` is:
```
function isassigned(a::AbstractArray, i::Integer...)
    try
        a[i...]
        true
    catch e
        if isa(e, BoundsError) || isa(e, UndefRefError)
            return false
        else
            rethrow()
        end
    end
end
```
and for `Array` is:
```
function isassigned(a::Array, i::Int...)
    @_inline_meta
    ii = (_sub2ind(size(a), i...) % UInt) - 1
    @boundscheck ii < length(a) % UInt || return false
    ccall(:jl_array_isassigned, Cint, (Any, UInt), a, ii) == 1
end
```
and this is obviously faster.

> Using a table with column access, it took 0.001s (mean) to print. Using a table with row access, it took 0.095s (mean).

This is my point - I think for printing this would be almost non-noticeable. Also I assume you do it on Linux terminal where the difference is most noticeable. On Windows or in Jupyter Notebook it would be less visible.

**Comment by ronisbr** [comment]
> @ronisbr Maybe the performance difference is due to the fact that `isassigned` is in a separate function? Adding `try... catch` may disable inlining. You can try with an explicit `@inline`.

I added `@inline` to those functions and I still get the same timings. I really have no idea, maybe the other version is using the faster `isassigned` for arrays.

> This is my point - I think for printing this would be almost non-noticeable. Also I assume you do it on Linux terminal where the difference is most noticeable. On Windows or in Jupyter Notebook it would be less visible.

I totally agree! I have already pushed the new commit. PrettyTables.jl now has support for `#undef` for any kind of Tables.jl (row or column accesses). I was justing wondering if I am doing something really bad :D
